text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+    }
+
+Introduce the abstract classes for queries:
+
+.. code-block:: python
+
+    class AsyncExecutor(abc.AbstraceBaseClass):
+        async def execute(self, query): ...
+        async def query(self, query: str, *args, **kwargs) -> datatypes.Set: ...
+        async def query_one(self, query: str, *args, **kwargs) -> typing.Any: ...
+        async def query_json(self, query: str, *args, **kwargs) -> str: ...
+        async def query_one_json(self, query: str, *args, **kwargs) -> str: ...
+
+    class Executor(abc.AbstractClass):
+        def query(self, query: str, *args, **kwargs) -> datatypes.Set: ...
+        def query_one(self, query: str, *args, **kwargs) -> typing.Any: ...
+        def query_json(self, query: str, *args, **kwargs) -> str: ...
+        def query_one_json(self, query: str, *args, **kwargs) -> str: ...
+        def execute(self, query: str) -> None: ...
+
+These base classes should be implemented by respective classes:
+
+.. code-block:: python
+
+    class AsyncIOTransaction(AsyncExecutor): ...
+    class AsyncIOConnection(AsyncExecutor): ...
+    class AsyncIOPool(AsyncExecutor): ...
+    class Transaction(Executor): ...
+    class Connection(Executor): ...
+    class Pool(Executor): ...
+
+
+Exceptions API
+``````````````
+
+Error hierarchy is amended by introducing ``TransientError``, ``NetworkError``
+and ``EarlyNetworkError`` with the following relationships:
+
+.. code-block:: python
+
+    class TransientError(TransactionError): ...
+    class TransactionDeadlockError(TransientError): ...
+    class TransactionSerializationError(TransientError): ...
+    class NetworkError(EdgeDBError): ...
+    class EarlyNetworkError(NetworkError): ...
+
+All network error within connection should be converted into
+``EarlyNetworkError`` or ``NetworkError``. Former is used in context
+where we catch network error before sending a request.
+
+
+Future Work
+===========
+
+Transaction on Specific Connection
+----------------------------------
+
+Do we want and how ``connection.retry()`` should work?
+
+a. It may only retry on the same connection and fail on disconnect
+b. It may reconnect and replace underlying socket in the connection
+   object and retry
+c. We may only allow transactions on connection pools
+
+This can be postponed to later RFC when we know what are use cases of
+it.
+
+
+Retry Single Queries
+--------------------
+
+While it's tempting to retry ``pool.query`` and ``pool.execute`` calls,
+it **gives the false sense of security**: no "concurrent update" issues
+seen.  But it's better to see such error and turn the whole block of
+code into a transaction rather than just a mutation. I.e. retrying a
+single mutable request on a "concurrent update" error must be a
+deliberate decision.
+
+
+Retry Single Queries on Connection Error
+----------------------------------------
+
+To make changing EdgeDB address or restarting EdgeDB work nicely, we
+need to retry simple queries on connection errors too:
```

Comment:
"we need to retry simple queries on connection errors too" this is very questionable (and it looks like it can only apply to the pooling APIs specifically, right?)

IMO, crashing on `ConnectionError` is the only sane option, and those who want to make their code more resilient can create a wrapper API.

Reply:
What's questionable?
```python
def product_index(req):
    return template(self.db.query("SELECT Product{...}"))

def buy_product(req):
    for tx in self.db.retry():
        with tx:
            price = db.query_row("SELECT price := Product.price...")
            db.query_row("INSERT ShopingCart {...}", ...)
```
Now if we restart server, or change address, or do any other networking changes (i.e. in docker you can get a network error even if just changing unrelated container), user can press "buy" button without seeing error page, but can't get a read-only index page without hitting `500 Internal Server Error`.

What we optimize for? Do we recommend read-only transactions? Do we recommend "roll your own wrapper"?